### PR TITLE
Update `add_purchase_order` operation

### PR DESCRIPTION
**Pull request by dnewh**
This updates the `add_purchase_order` operation to properly add/update POs, versions, and revisions while avoiding unnecessary db updates.

### CODE REVIEW

**Comment by dnewh** [comment]
depends on PR #932

File: sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                    version_query =
+                        version_query.filter(purchase_order_version::service_id.is_null());
+                }
+
+                let version = version_query
+                    .first::<PurchaseOrderVersionModel>(conn)
+                    .map(Some)
+                    .map_err(|err| {
+                        if err == dsl_error::NotFound {
+                            PurchaseOrderStoreError::ConstraintViolationError(
```
Comment (shannynalayna) [inline sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs:100]: Seems like the `NotFound` variant would fit better here?

File: sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+                query = query.filter(purchase_order_version_revision::service_id.eq(service_id));
+            } else {
+                query = query.filter(purchase_order_version_revision::service_id.is_null());
+            }
+
+            let duplicate = query
+                .first::<PurchaseOrderVersionRevisionModel>(conn)
+                .map(Some)
+                .or_else(|err| {
+                    if err == dsl_error::NotFound {
```
Comment (shannynalayna) [inline sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs:200]: `.optional()` could be used here too

File: sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                    version_query =
+                        version_query.filter(purchase_order_version::service_id.is_null());
+                }
+
+                let version = version_query
+                    .first::<PurchaseOrderVersionModel>(conn)
+                    .map(Some)
+                    .map_err(|err| {
+                        if err == dsl_error::NotFound {
+                            PurchaseOrderStoreError::ConstraintViolationError(
```
Comment (shannynalayna) [inline sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs:239]: `NotFound` variant thing here too

File: sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs (hunk 2 above)
Comment (shannynalayna) [reply sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version_revision.rs:200]: Across all instances of this or_else in the ops you're adding, really

File: sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                        .filter(
+                            purchase_order_version::purchase_order_uid
+                                .eq(&version.purchase_order_uid)
+                                .and(purchase_order_version::version_id.eq(&version.version_id))
+                                .and(purchase_order_version::is_draft.eq(&version.is_draft))
+                                .and(
+                                    purchase_order_version::current_revision_id
+                                        .eq(&version.current_revision_id),
+                                )
+                                .and(purchase_order_version::service_id.eq(service_id)) //Update here
```
Comment (shannynalayna) [inline sdk/src/purchase_order/store/diesel/operations/add_purchase_order_version.rs:85]: could these comments maybe be removed or updated to clarify the meaning ? Same with below. Kinda confused what they're supposed to mean

**Review by shannynalayna — APPROVED** [review]

**Review by leebradley — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**